### PR TITLE
core: do not drop backslashes in SYSTEMD_WANTS=

### LINKED ISSUE #16735: SYSTEMD_WANTS does not preserve backslashes
**systemd version the issue has been seen with**
241-7~deb10u3 (but this is likely applicable to `master` as well)

**Used distribution**
ev3dev-buster (~= Debian buster)

**Expected behaviour you didn't see**
Backslashes are preserved in unit names coming from udev `SYSTEMD_WANTS`. The following udev rule should start a unit named `ev3-uart@ev3\x2dports:in4.service` if `LEGO_ADDRESS` is `ev3-ports:in4`:
```
SUBSYSTEMS=="lego", ENV{DEVTYPE}=="ev3-uart-host", ACTION=="add", TAG+="systemd", \
    PROGRAM="/bin/systemd-escape $env{LEGO_ADDRESS}" \
    ENV{SYSTEMD_WANTS}="ev3-uart@$result.service", \
    ENV{SYSTEMD_ALIAS}="/sys/subsystem/lego/devices/ev3_uart_host_$env{LEGO_ADDRESS}"
```

**Unexpected behaviour you saw**
Backslashes are stripped, therefore breaking manual escaping. Unit named `ev3-uart@ev3x2dports:in4.service` is started instead of the one above. The template unit is then unable to unescape the `ev3x2dports:in4` instance name. However, the unit requires both the escaped and unescaped names:
```
# This file is part of the ev3-systemd package

[Unit]
Description=LEGO MINDSTORMS EV3 UART Sensors
# This is the alias created by the ev3dev-uart udev rule.
BindsTo=sys-subsystem-lego-devices-ev3_uart_host_%i.device
Wants=systemd-udev-settle.service
After=systemd-udev-settle.service

[Service]
Type=forking
ExecStart=/usr/sbin/ldattach 29 /dev/tty_%I
```

I think the problem is in https://github.com/systemd/systemd/blob/61c8b73a35fd71cb7cbb41b78d93a124bf347e8e/src/core/device.c#L379, where the ~quote~ backslash may be stripped as part of unquoting.

Hardcoding `SYSTEMD_WANTS` to contain a second backslash does work around the problem, but this is more difficult to do automatically inside the udev rule.

**Steps to reproduce the problem**
 * Find a device that can be hot-swapped
 * Create a template systemd unit that depends on both `%i` and `%I` and accepts a non-sysfs escaped string as an instance name
 * Add a udev rule for the device with a `SYSTEMD_WANTS` property that either uses `systemd-escape` result or hardcodes the instance name with only one slash
 * Observe that the unit started on device hotplug is not able to reverse the escaping done by `systemd-escape`

**Related bugs**
OP in https://github.com/systemd/systemd/issues/7109 observed something similar. Using empty instance name as introduced in https://github.com/systemd/systemd/commit/dcebc9bae4dcc3e844f01558c6127fc0d8745c8e would be another way to work around this, albeit not an ideal one.

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #16735.

### CODE REVIEW

**Comment by TriMoon** [comment]
@yuwata  You might want to update/remove this PR also, see the discussion results of #16735  :wink:

File: src/core/device.c [hunk 1]
```diff
@@ -376,7 +376,7 @@ static int device_add_udev_wants(Unit *u, sd_device *dev) {
         for (;;) {
                 _cleanup_free_ char *word = NULL, *k = NULL;
 
-                r = extract_first_word(&wants, &word, NULL, EXTRACT_UNQUOTE);
+                r = extract_first_word(&wants, &word, NULL, EXTRACT_UNQUOTE | EXTRACT_RETAIN_ESCAPE);
```
Comment (keszybz) [inline src/core/device.c:379]: This doesn't look right.

**Comment by yuwata** [comment]
> I agree with @TriMoon and @JakubVanek that this doesn't seem to be the right thing to do. But the first commit is OK. Can you drop the second one?

Hmm. This PR slightly(?) breaks backward compatibility. Is it the reason this PR is not good? But parsing without `EXTRACT_RETAIN_ESCAPE` makes `systemd-escape` almost unusable in udev rules... And as described in #16760, currently we do not have a way to double escape string by using "PROGRAM=" key... Of course, adding a new key or option can solve the issue explained in #16760. But it looks not elegant for me... Or am I missed some point discussed in #16735?

**Comment by TriMoon** [comment]
>  @yuwata 
> Or am I missed some point discussed in #16735?

Maybe: https://github.com/systemd/systemd/issues/16735#issuecomment-674791762 ?

**Comment by keszybz** [comment]
For me, EXTRACT_RETAIN_ESCAPE is a work-around. We effectively do unescaping in two steps: first some characters, then other characters. That is why my initial reaction was to say that this is not a valid approach. It would help a lot if you could describe *how* this is supposed to work: what happens in various cases with this patch applied. Then we can talk more reasonably whether this is the right approach.

File: src/core/device.c [hunk 2]
```diff
@@ -441,10 +441,7 @@ static int device_add_udev_wants(Unit *u, sd_device *dev) {
                 }
         }
 
-        strv_free(d->wants_property);
-        d->wants_property = TAKE_PTR(added);
-
-        return 0;
+        return strv_free_and_replace(d->wants_property, added);
```
Comment (poettering) [inline src/core/device.c]: maybe split this out into its own PR so tht we can merge this before this PR is settled? this looks uncontroversial after all

Comment (yuwata) [reply src/core/device.c]: Moved to #17057.

**Comment by yuwata** [comment]
Hmm, I noticed that it seems not necessary to escape the instance string for the case #16737, as the specified unit name in `SYSTEMD_WANTS=` will be escaped: https://github.com/systemd/systemd/blob/67f5ae2d6986d0271fa6223a68fe0104957de863/src/core/device.c#L403

Closing, as now I am very confused by the original issue...

**Comment by yuwata** [comment]
This is not merged.

**Comment by CarloWood** [comment]
Hello all. https://github.com/systemd/systemd/issues/16735#issuecomment-2561123126 got just re-opened. This issue still exists and simply was never addressed.

Can a maintainer please confirm the bug by executing the test case that I posted there (see above link)?

**Comment by JakubVanek** [comment]
Hello, I have modified @CarloWood's test case (BTW: thank you! having it available helped me a lot) to show that this PR actually restores the behaviour of systemd <=v232 (last release before commit https://github.com/systemd/systemd/commit/ceed8f0c8b9a46300eccd1afa2dd8d3c2cb6b47c). See https://github.com/systemd/systemd/issues/16735#issuecomment-2563001829 for reproduction steps, and the comments below for ongoing discussion.

I think some further investigation may be needed to check that this PR handles all edge cases. However, for the basic instance name escaping in the original issue it seems to work OK.

**Comment by JakubVanek** [comment]
The `EXTRACT_RETAIN_ESCAPE` does work a bit differently after all. I have extracted both functions into small programs and compared the words they produce:

* `FOREACH_WORD_QUOTED()` is the systemd v232 behaviour
* `extract_first_word(..., EXTRACT_UNQUOTE)` is the current systemd behaviour
* `extract_first_word(..., EXTRACT_UNQUOTE \| EXTRACT_RETAIN_ESCAPE)` is the systemd behaviour after applying this PR


| `SYSTEMD_WANTS` value | `FOREACH_WORD_QUOTED()` | `extract_first_word(..., EXTRACT_UNQUOTE)` | `extract_first_word(..., EXTRACT_UNQUOTE \| EXTRACT_RETAIN_ESCAPE)` |
|-----------------------|-------------------------|--------------------------------------------|---------------------------------------------------------------------|
|          `a`          |           `a`           |                     `a`                    |                                 `a`                                 |
| `"a"`                 | `a`                     | `a`                                        | `a`                                                                 |
| `a b`                 | `a`, `b`                | `a`, `b`                                   | `a`, `b`                                                            |
| `"a" "b"`             | `a`, `b`                | `a`, `b`                                   | `a`, `b`                                                            |
| `"a\"a"`              | `a\"a`                  | `a"a`                                      | error -22                                                           |
| `a\ b`                | `a\ b`                  | `a b`                                      | `a\`, `b`                                                           |
| `"a a"`               | `a a`                   | `a a`                                      | `a a`                                                               |
| `"a a" b`             | `a a`, `b`              | `a a`, `b`                                 | `a a`, `b`                                                          |
| `"a\ b" c`            | `a\ b`, `c`             | `a b`, `c`                                 | `a\ b`, `c`                                                         |
| `"a`                  | nothing                 | error -22                                  | error -22                                                           |
| `"a\`                 | nothing                 | error -22                                  | error -22                                                           |
| `a\`                  | nothing                 | error -22                                  | `a\`                                                                |
| `a\ `                 | `a\ `                   | `a `                                       | `a\`                                                                |
| `"a"b`                | nothing                 | `ab`                                       | `ab`                                                                |

Notably, this PR would change the parsing of `a\ b`.

<details>
<summary>C program for reproducing the systemd v232 SYSTEMD_WANTS parsing behaviour</summary>

Compile simply using
```bash
gcc sd_v232_parsing.c -o sd_v232_parsing
```

The program expects the `SYSTEMD_WANTS` patterns from udev as its arguments:
```
$ ./sd_v232_parsing 'printer@triggered\x2dby\x2dudev.service'
* `printer@triggered\x2dby\x2dudev.service` -> `printer@triggered\x2dby\x2dudev.service`
```

```c
#include <stdio.h>
#include <string.h>
#include <stdbool.h>
#include <assert.h>

#define WHITESPACE        " \t\n\r"

static size_t strcspn_escaped(const char *s, const char *reject) {
        bool escaped = false;
        int n;

        for (n=0; s[n]; n++) {
                if (escaped)
                        escaped = false;
                else if (s[n] == '\\')
                        escaped = true;
                else if (strchr(reject, s[n]))
                        break;
        }

        /* if s ends in \, return index of previous char */
        return n - escaped;
}

/* Split a string into words. */
const char* split(const char **state, size_t *l, const char *separator, bool quoted) {
        const char *current;

        current = *state;

        if (!*current) {
                assert(**state == '\0');
                return NULL;
        }

        current += strspn(current, separator);
        if (!*current) {
                *state = current;
                return NULL;
        }

        if (quoted && strchr("\'\"", *current)) {
                char quotechars[2] = {*current, '\0'};

                *l = strcspn_escaped(current + 1, quotechars);
                if (current[*l + 1] == '\0' || current[*l + 1] != quotechars[0] ||
                    (current[*l + 2] && !strchr(separator, current[*l + 2]))) {
                        /* right quote missing or garbage at the end */
                        *state = current;
                        return NULL;
                }
                *state = current++ + *l + 2;
        } else if (quoted) {
                *l = strcspn_escaped(current, separator);
                if (current[*l] && !strchr(separator, current[*l])) {
                        /* unfinished escape */
                        *state = current;
                        return NULL;
                }
                *state = current + *l;
        } else {
                *l = strcspn(current, separator);
                *state = current + *l;
        }

        return current;
}

#define FOREACH_WORD(word, length, s, state)                            \
        _FOREACH_WORD(word, length, s, WHITESPACE, false, state)

#define FOREACH_WORD_SEPARATOR(word, length, s, separator, state)       \
        _FOREACH_WORD(word, length, s, separator, false, state)

#define FOREACH_WORD_QUOTED(word, length, s, state)                     \
        _FOREACH_WORD(word, length, s, WHITESPACE, true, state)

#define _FOREACH_WORD(word, length, s, separator, quoted, state)        \
        for ((state) = (s), (word) = split(&(state), &(length), (separator), (quoted)); (word); (word) = split(&(state), &(length), (separator), (quoted)))


int main(int argc, char **argv) {
    for (int arg = 1; arg < argc; arg++) {
        const char *wants;
        const char *word, *state;
        size_t l;

        wants = argv[arg];
        printf("* `%s` -> ", wants);
        int count = 0;
        bool first = true;
        FOREACH_WORD_QUOTED(word, l, wants, state) {
            if (first) {
                first = false;
            } else {
                printf(", ");
            }
            printf("`%.*s`", (int) l, word);
            count++;
        }
        if (count == 0) {
            printf("null / end-of-iteration");
        }
        printf("\n");
    }
}
```

</details>

<details>
<summary>C program for reproducing the current systemd SYSTEMD_WANTS parsing behaviour</summary>

Compile simply using
```bash
gcc sd_current_parsing.c -o sd_current_parsing
```

The program expects the `SYSTEMD_WANTS` patterns from udev as its arguments:
```
$ ./sd_current_parsing 'printer@triggered\x2dby\x2dudev.service'
* `printer@triggered\x2dby\x2dudev.service` -> `printer@triggered\x2dby\x2dudev.service`
```

To switch between the versions with and without `EXTRACT_RETAIN_ESCAPE`,
change the `flags` variable in the `main` function and recompile the program.

```c
#include <string.h>
#include <errno.h>
#include <stdarg.h>
#include <stdbool.h>
#include <stddef.h>
#include <stdint.h>
#include <stdlib.h>
#include <assert.h>
#include <uchar.h>
#include <memory.h>
#include <malloc.h>

static inline void freep(void *p) {
        free(*(void **) p);
}

#define _cleanup_free_ _cleanup_(freep)
#define _cleanup_(x) __attribute__((__cleanup__(x)))
#define _unused_ __attribute__((__unused__))
#define _unlikely_(x) (__builtin_expect(!!(x), 0))
#define assert_not_reached() __builtin_unreachable()

#define FLAGS_SET(v, flags) \
        ((~(v) & (flags)) == 0)

#define GREEDY_REALLOC(array, need)                                     \
        greedy_realloc((void**) &(array), (need), sizeof((array)[0]))

#define MALLOC_SIZEOF_SAFE(x) \
        malloc_sizeof_safe((void**) &__builtin_choose_expr(__builtin_constant_p(x), (void*) { NULL }, (x)))

#define MUL_SAFE(ret, a, b) (!__builtin_mul_overflow(a, b, ret))
#define MUL_ASSIGN_SAFE(a, b) __MUL_ASSIGN_SAFE(UNIQ, a, b)
#define __MUL_ASSIGN_SAFE(q, a, b)                              \
        ({                                                      \
                const typeof(a) UNIQ_T(A, q) = (a);             \
                MUL_SAFE(UNIQ_T(A, q), *UNIQ_T(A, q), b);       \
        })

#define UNIQ_T(x, uniq) CONCATENATE(__unique_prefix_, CONCATENATE(x, uniq))
#define UNIQ __COUNTER__

#define XCONCATENATE(x, y) x ## y
#define CONCATENATE(x, y) XCONCATENATE(x, y)

#define TAKE_GENERIC(var, type, nullvalue)                       \
        ({                                                       \
                type *_pvar_ = &(var);                           \
                type _var_ = *_pvar_;                            \
                type _nullvalue_ = nullvalue;                    \
                *_pvar_ = _nullvalue_;                           \
                _var_;                                           \
        })
#define TAKE_PTR_TYPE(ptr, type) TAKE_GENERIC(ptr, type, NULL)
#define TAKE_PTR(ptr) TAKE_PTR_TYPE(ptr, typeof(ptr))

int cunescape_one(const char *p, size_t length, char32_t *ret, bool *eight_bit, bool accept_nul) {
    return -EINVAL;
}

size_t utf8_encode_unichar(char *out_utf8, char32_t g) {
    return 0;
}

void *expand_to_usable(void *ptr, size_t newsize _unused_) {
        return ptr;
}

static inline size_t malloc_sizeof_safe(void **xp) {
        if (_unlikely_(!xp || !*xp))
                return 0;

        size_t sz = malloc_usable_size(*xp);
        *xp = expand_to_usable(*xp, sz);
        /* GCC doesn't see the _returns_nonnull_ when built with ubsan, so yet another hint to make it doubly
         * clear that expand_to_usable won't return NULL.
         * See: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=79265 */
        if (!*xp)
                assert_not_reached();
        return sz;
}

void* greedy_realloc(
                void **p,
                size_t need,
                size_t size) {

        size_t newalloc;
        void *q;

        assert(p);

        /* We use malloc_usable_size() for determining the current allocated size. On all systems we care
         * about this should be safe to rely on. Should there ever arise the need to avoid relying on this we
         * can instead locally fall back to realloc() on every call, rounded up to the next exponent of 2 or
         * so. */

        if (*p && (size == 0 || (MALLOC_SIZEOF_SAFE(*p) / size >= need)))
                return *p;

        if (_unlikely_(need > SIZE_MAX/2)) /* Overflow check */
                return NULL;
        newalloc = need * 2;

        if (!MUL_ASSIGN_SAFE(&newalloc, size))
                return NULL;

        if (newalloc < 64) /* Allocate at least 64 bytes */
                newalloc = 64;

        q = realloc(*p, newalloc);
        if (!q)
                return NULL;

        return *p = q;
}

#define WHITESPACE        " \t\n\r"

typedef enum ExtractFlags {
        EXTRACT_RELAX                    = 1 << 0, /* Allow unbalanced quote and eat up trailing backslash. */
        EXTRACT_CUNESCAPE                = 1 << 1, /* Unescape known escape sequences. */
        EXTRACT_UNESCAPE_RELAX           = 1 << 2, /* Allow and keep unknown escape sequences, allow and keep trailing backslash. */
        EXTRACT_UNESCAPE_SEPARATORS      = 1 << 3, /* Unescape separators (those specified, or whitespace by default). */
        EXTRACT_KEEP_QUOTE               = 1 << 4, /* Ignore separators in quoting with "" and ''. */
        EXTRACT_UNQUOTE                  = 1 << 5, /* Ignore separators in quoting with "" and '', and remove the quotes. */
        EXTRACT_DONT_COALESCE_SEPARATORS = 1 << 6, /* Don't treat multiple adjacent separators as one */
        EXTRACT_RETAIN_ESCAPE            = 1 << 7, /* Treat escape character '\' as any other character without special meaning */
        EXTRACT_RETAIN_SEPARATORS        = 1 << 8, /* Do not advance the original string pointer past the separator(s) */

        /* Note that if no flags are specified, escaped escape characters will be silently stripped. */
} ExtractFlags;



int extract_first_word(const char **p, char **ret, const char *separators, ExtractFlags flags) {
        _cleanup_free_ char *s = NULL;
        size_t sz = 0;
        char quote = 0;                 /* 0 or ' or " */
        bool backslash = false;         /* whether we've just seen a backslash */
        char c;
        int r;

        assert(p);
        assert(ret);
        assert(!FLAGS_SET(flags, EXTRACT_KEEP_QUOTE | EXTRACT_UNQUOTE));

        /* Bail early if called after last value or with no input */
        if (!*p)
                goto finish;
        c = **p;

        if (!separators)
                separators = WHITESPACE;

        /* Parses the first word of a string, and returns it in
         * *ret. Removes all quotes in the process. When parsing fails
         * (because of an uneven number of quotes or similar), leaves
         * the pointer *p at the first invalid character. */

        if (flags & EXTRACT_DONT_COALESCE_SEPARATORS)
                if (!GREEDY_REALLOC(s, sz+1))
                        return -ENOMEM;

        for (;; (*p)++, c = **p) {
                if (c == 0)
                        goto finish_force_terminate;
                else if (strchr(separators, c)) {
                        if (flags & EXTRACT_DONT_COALESCE_SEPARATORS) {
                                if (!(flags & EXTRACT_RETAIN_SEPARATORS))
                                        (*p)++;
                                goto finish_force_next;
                        }
                } else {
                        /* We found a non-blank character, so we will always
                         * want to return a string (even if it is empty),
                         * allocate it here. */
                        if (!GREEDY_REALLOC(s, sz+1))
                                return -ENOMEM;
                        break;
                }
        }

        for (;; (*p)++, c = **p) {
                if (backslash) {
                        if (!GREEDY_REALLOC(s, sz+7))
                                return -ENOMEM;

                        if (c == 0) {
                                if ((flags & EXTRACT_UNESCAPE_RELAX) &&
                                    (quote == 0 || flags & EXTRACT_RELAX)) {
                                        /* If we find an unquoted trailing backslash and we're in
                                         * EXTRACT_UNESCAPE_RELAX mode, keep it verbatim in the
                                         * output.
                                         *
                                         * Unbalanced quotes will only be allowed in EXTRACT_RELAX
                                         * mode, EXTRACT_UNESCAPE_RELAX mode does not allow them.
                                         */
                                        s[sz++] = '\\';
                                        goto finish_force_terminate;
                                }
                                if (flags & EXTRACT_RELAX)
                                        goto finish_force_terminate;
                                return -EINVAL;
                        }

                        if (flags & (EXTRACT_CUNESCAPE|EXTRACT_UNESCAPE_SEPARATORS)) {
                                bool eight_bit = false;
                                char32_t u;

                                if ((flags & EXTRACT_CUNESCAPE) &&
                                    (r = cunescape_one(*p, SIZE_MAX, &u, &eight_bit, false)) >= 0) {
                                        /* A valid escaped sequence */
                                        assert(r >= 1);

                                        (*p) += r - 1;

                                        if (eight_bit)
                                                s[sz++] = u;
                                        else
                                                sz += utf8_encode_unichar(s + sz, u);
                                } else if ((flags & EXTRACT_UNESCAPE_SEPARATORS) &&
                                           (strchr(separators, **p) || **p == '\\'))
                                        /* An escaped separator char or the escape char itself */
                                        s[sz++] = c;
                                else if (flags & EXTRACT_UNESCAPE_RELAX) {
                                        s[sz++] = '\\';
                                        s[sz++] = c;
                                } else
                                        return -EINVAL;
                        } else
                                s[sz++] = c;

                        backslash = false;

                } else if (quote != 0) {     /* inside either single or double quotes */
                        for (;; (*p)++, c = **p) {
                                if (c == 0) {
                                        if (flags & EXTRACT_RELAX)
                                                goto finish_force_terminate;
                                        return -EINVAL;
                                } else if (c == quote) {        /* found the end quote */
                                        quote = 0;
                                        if (flags & EXTRACT_UNQUOTE)
                                                break;
                                } else if (c == '\\' && !(flags & EXTRACT_RETAIN_ESCAPE)) {
                                        backslash = true;
                                        break;
                                }

                                if (!GREEDY_REALLOC(s, sz+2))
                                        return -ENOMEM;

                                s[sz++] = c;

                                if (quote == 0)
                                        break;
                        }

                } else {
                        for (;; (*p)++, c = **p) {
                                if (c == 0)
                                        goto finish_force_terminate;
                                else if ((c == '\'' || c == '"') && (flags & (EXTRACT_KEEP_QUOTE | EXTRACT_UNQUOTE))) {
                                        quote = c;
                                        if (flags & EXTRACT_UNQUOTE)
                                                break;
                                } else if (c == '\\' && !(flags & EXTRACT_RETAIN_ESCAPE)) {
                                        backslash = true;
                                        break;
                                } else if (strchr(separators, c)) {
                                        if (flags & EXTRACT_DONT_COALESCE_SEPARATORS) {
                                                if (!(flags & EXTRACT_RETAIN_SEPARATORS))
                                                        (*p)++;
                                                goto finish_force_next;
                                        }
                                        if (!(flags & EXTRACT_RETAIN_SEPARATORS))
                                                /* Skip additional coalesced separators. */
                                                for (;; (*p)++, c = **p) {
                                                        if (c == 0)
                                                                goto finish_force_terminate;
                                                        if (!strchr(separators, c))
                                                                break;
                                                }
                                        goto finish;

                                }

                                if (!GREEDY_REALLOC(s, sz+2))
                                        return -ENOMEM;

                                s[sz++] = c;

                                if (quote != 0)
                                        break;
                        }
                }
        }

finish_force_terminate:
        *p = NULL;
finish:
        if (!s) {
                *p = NULL;
                *ret = NULL;
                return 0;
        }

finish_force_next:
        s[sz] = 0;
        *ret = TAKE_PTR(s);

        return 1;
}

int main(int argc, char **argv) {
    // ExtractFlags flags = EXTRACT_UNQUOTE;
    ExtractFlags flags = EXTRACT_UNQUOTE | EXTRACT_RETAIN_ESCAPE;

    for (int arg = 1; arg < argc; arg++) {
        _cleanup_free_ char *word = NULL, *k = NULL;
        const char *wants;


        wants = argv[arg];
        printf("* `%s` -> ", wants);

        int count = 0;
        for (;; count++) {
            _cleanup_free_ char *word = NULL, *k = NULL;
            int r = extract_first_word(&wants, &word, NULL, flags);
            if (r == 0) {
                break;
            }
            if (r < 0) {
                printf("error %d", r);
                break;
            }
            if (count > 0) {
                printf(", ");
            }
            printf("`%s`", word);
        }
        printf("\n");
    }
}
```

</details>

**Comment by arvidjaar** [comment]
> `a\`, `b`

Yes. `EXTRACT_RETAIN_ESCAPE` is misleading. It does not *retain* escape as `FOREACH_WORD_QUOTED` did - it effectively disables escaping by `\` turning it into an ordinal character. I have a nagging suspicion that it was by mistake in implementation.

I do not think it matters in real life for this particular use case though.

**Comment by JakubVanek** [comment]
I have created https://github.com/systemd/systemd/pull/35768 with an alternative solution that handles even the `a\ b` edge case. I realized a bit too late that the edge case is indeed not very useful (`a\ b` is not useful for escaping the whitespace). I wanted to make that code available anyway though.

Overall, I think that https://github.com/systemd/systemd/pull/16737 (e.g. the PR I'm writing this comment on) might be the better solution overall (less overall complexity).

**Comment by Olf0** [comment]
> `EXTRACT_RETAIN_ESCAPE` is misleading.

`EXTRACT_RETAIN_ESCAPE_ALT` as [suggested by @JakubVanek](https://github.com/systemd/systemd/pull/35768#discussion_r1898697472) is a bit better IMO.

WRT merging either #16737 or #35768, I do not really mind, but would also slightly tend to preferring  #16737.

**Comment by TriMoon** [comment]
> > `EXTRACT_RETAIN_ESCAPE` is misleading.
> 
> `EXTRACT_RETAIN_ESCAPE_ALT` as [suggested by @JakubVanek](https://github.com/systemd/systemd/pull/35768#discussion_r1898697472) is a bit better IMO.

`EXTRACT_RETAIN_BACKSLASH` https://github.com/systemd/systemd/pull/35768#discussion_r1899367064 is even better :wink:
